### PR TITLE
Fix deprecation warnings for each()

### PR DESCRIPTION
**Pull request by Nebel54**
As stated in #22 the each iterator is deprecated in PHP 7.2 and throws warnings. I used Ayesh's patch for Drupal 7 (https://www.drupal.org/project/drupal/issues/2925449) and backported it to D6.

### CODE REVIEW

File: includes/menu.inc [hunk 1]
```diff
@@ -1560,7 +1561,9 @@ function menu_set_active_trail($new_trail = NULL) {
     }
 
     $tree = menu_tree_page_data(menu_get_active_menu_name());
-    list($key, $curr) = each($tree);
+
+    $curr = current($tree);
+    next($tree);
```
Comment (dsnopek) [inline includes/menu.inc:1566]: Couldn't this just be one line?

```
$curr = next($tree);
```

**Comment by dsnopek** [comment]
Thanks for the PR! I've been putting off dealing with `each()` because it's only deprecated, but will still work in PHP 7.2. Once we're passed the 5.6 and 7.0 EOL, I'll probably dig deeper into this.

File: includes/menu.inc (hunk 1 above)
Comment (Nebel54) [reply includes/menu.inc:1566]: No, I don't think so - both variables ($curr and $tree) are used in the while clause after this code.

Comment (rmuit) [reply includes/menu.inc:1566]: The code is indeed correct as-is. This is the literal conversion of what each() does: _first_ assign the current value, _then_ advance $tree. This means "$curr = next($tree);" would not be correct, because the while($curr) loop wants to process the _current_ value, which $tree has already advanced beyond.

To say this differently: the while($curr) loop is adjusted to run correctly with this somewhat strange code, and you _can_ make this one line, but that _also_ means you have to adjust the while loop. The one line would be:
```
$curr = current($tree);
``` 
...and then you would have to add a "} else {" to the while loop, as in:
```
    $tree = $curr['below'] ? $curr['below'] : array(); // <<<<<< This code exists; see below
  }
  else {
    next($tree);
  }
  $curr = current($tree);   //// <<<< now the next() is not necessary here
```

**My suggestion:** Either do what is described in this comment, or put a comment above the above next() statement, because it's indeed hard to decipher this:
```
// The while() loop assumes that the $tree array pointer advances already.
```

File: modules/book/book.module [hunk 2]
```diff
@@ -545,11 +545,13 @@ function book_prev($book_link) {
     return NULL;
   }
   $flat = book_get_flat_menu($book_link);
-  // Assigning the array to $flat resets the array pointer for use with each().
+  reset($flat);
```
Comment (rmuit) [inline modules/book/book.module:548]: This is interesting. I don't think this is necessary, since the comment which you're removing explicitly states that it isn't. (And the example in https://www.php.net/manual/en/function.current.php) also implies that it isn't.) The only thing that would need to be changed is strip the " for use with each()" off the end of the comment.

I can totally understand writing explicit code like reset() calls, so we're not dependent on half-documented internal PHP features. I do it myself too.
...But... I'm still commenting because I notice the PR is inconsistent :) No reset() was added in the above menu.inc code either. That may be a reason to not do it here either.

**Comment by rmuit** [comment]
Thanks for the patch @Nebel54.

I commented with nitpicks on the PR as a sign that it's been reviewed. And it's being used in other D6 projects. But
* The nitpicks are really nitpicks (though I'm still a fan of at least adding one comment line); the code is good as is
* And now afterwards, I see that this code I'm picking apart, is already used in D7.

So that can be a reason to just include it.

Alternatively, I'd be happy to do a tweaked PR with the nitpicks included... let me know.

**Comment by pentium10** [comment]
this should be merged

File: modules/book/book.module (hunk 2 above)
Comment (rmuit) [reply modules/book/book.module:548]: I read the D7 patch again, and
* andypost explicitly argues for including a reset(), in https://www.drupal.org/project/drupal/issues/2925449#comment-12380008. \*shrugs\* 
* then says "other places are still missing it too" in # 21, which leads to a boatload of confusion and wrong patches in the rest of the issue because people are reset()s in the wrong places.

Sigh. Never mind, let's not be consistent, let's just commit the resulting D7 patch version.

**Comment by rmuit** [comment]
Yes, this should be merged.

I'm sorry for not seeing that this was a direct copy of a D7 patch, and being too wordy in my review. I should just have answered @dsnopek "I agree, the code is indeed correct as-is." and left it at that.

The code isn't the most nicely readable but IMHO we should keep the D6 code changes small, especially if that keeps the code equal to the D7 version. (Even if https://github.com/kaznovac/drupal/commit/e0874b36fce796efff24396e1d012234c794212d results in nicer more readable code.)

I went over the whole thing again, and it's fine. The only thing I would personally do is include just one comment line as per https://github.com/rmuit/drupal/commit/b8ffee8bc2c76015c634dfec3de4809f357df845 (which is rmuit:php-7, a commit directly on top of this PR) but that's completely optional.

**Comment by dsnopek** [comment]
Thanks, Everyone! Finally got back around to reviewing and merging this :-)
